src/nvim/eval.c

```diff
@@ -16753,6 +16855,83 @@ static void f_visualmode(typval_T *argvars, typval_T *rettv)
     curbuf->b_visual_mode_eval = NUL;
 }
 
+// watcheradd(dict, key, funcref) function
+static void f_watcheradd(typval_T *argvars, typval_T *rettv)
+{
+  if (check_restricted() || check_secure()) {
+    return;
+  }
+
+  if (argvars[0].v_type != VAR_DICT ||
+      argvars[1].v_type != VAR_STRING ||
+      (argvars[2].v_type != VAR_FUNC && argvars[2].v_type != VAR_STRING)) {
+    // Wrong argument types
+    EMSG(_(e_invarg));
```

Comment:
It would be much better to report something more specific: at least, argument number which has invalid type.